File: cassidoo-newsletter/#311/cpp/anagram.cpp
// Given two strings s and t, return true if t is an anagram of s,
// and false otherwise.
#include <string>
#include <unordered_map>

bool isAnagram(std::string s, std::string t) {
  if (s.length() != t.length()) {
    return false;
  }
  std::unordered_map<char, int> histogram;
  for (int i = 0; i < s.size(); i++) {
    histogram[s[i]]++;
    histogram[t[i]]--;
  }
  for (auto it : histogram) {
    if (it.second)
      return false;
  }
  return true;
}

File: cassidoo-newsletter/#311/java/Anagram.java
// given two strings `s` and `t` returns `true` if they're anagrams,
// false otherwise

class Anagram {
    private static int HISTOGRAM_SIZE = 256;

    public boolean isAnagram(String s, String t) {
        if(s.length() != t.length()) return false;

        int histogram[] = new int[HISTOGRAM_SIZE];
        for(int i = 0; i < s.length(); i++) {
            histogram[s.charAt(i)]++;
            histogram[t.charAt(i)]--;
        }
        // valid anagram should have the `histogram` filled with `0s`:-)
        for(int e: histogram) {
            if (e != 0) return false;
        }
        return true;
    }
}
